File: Disjoint_Sets/Disjoint_Sets.py
class disjointSets:

    def __init__(self,num):
        self.N=num
        self.rank=[0]*(num+1)
        self.parent=[0]*(num+1)
        for i in range(num+1):
            self.parent[i]=i

    def find(self,x):
        if(self.parent[x]!=x):
            self.parent[x]=self.find(self.parent[x])
        return self.parent[x]

    def union(self,x,y):
        xr=self.find(x)
        yr=self.find(y)
        if(xr!=yr):
            if(self.rank[xr]<self.rank[yr]):

                self.parent[xr]=yr

            elif(self.rank[yr]<self.rank[xr]):

                self.parent[yr]=xr

            else:
                
                self.parent[yr]=xr
                self.rank[xr]=self.rank[xr]+1

# Example graph with two disjoint sets
''' 
1---2---3     6--4---5
'''
#number of nodes in the graph
n = 6
#initialize the disjoint set for n nodes   
d_set = disjointSets(n)
#number of edges
k = 4
edges=[]
edges.append([1,2]) #edge between 1 and 2
edges.append([2,3]) #edge between 2 and 3
edges.append([4,5]) #edge between 4 and 5
edges.append([4,6]) #edge between 4 and 6
edges.sort()
for i in range(k):
    d_set.union(edges[i][0],edges[i][1])

# check if 4 and 6 are in the same set
if(d_set.find(4)==d_set.find(6)):
    print("YES")
else:
    print("NO")
print("")
# check if 1 and 4 are in the same set
if(d_set.find(1)==d_set.find(4)):
    print("YES")
else:
    print("NO")
print("")

File: Disjoint_Sets/Disjoint_Sets.java
import java.util.Scanner;

public class Disjoint_Sets {
    static int[] parent;
    static int[] size;
    
    // function that initially defines the parent of all element as itself.
    static void makeSet(int n) {
        for (int i = 0; i < n + 1; i++) {
            parent[i] = i;
            size[i] = 1;
        }
        return;
    }
    
    // recursive function to find the parent of an element.
    static int find(int a) {
        if (parent[a] == a) return a;
        else return find(parent[a]);        
    }
    
    // function to merge two elements such that both of them belong to the same set
    static void union(int a, int b) {
        int parent_a = find(a);
        int parent_b = find(b);
        
        if (parent_a != parent_b) {
            if (size[parent_a] < size[parent_b]) {
                
                // swap a and b
                int temp = a;
                a = b;
                b = temp;
            }
            parent[parent_b] = parent_a;
            size[a] += size[b];
        }
    }
    
    public static void main(String[] args) {
        Scanner input = new Scanner(System.in);
        System.out.println("Enter number of nodes and number of edges");
        int n = input.nextInt();
        int e = input.nextInt();
        parent = new int[n + 1];
        size = new int[n + 1];
        System.out.println("Enter e edges");

        // make set with n nodes (1,2,...,n)
        makeSet(n);     
        for (int i = 0; i < e; i++) {
            int a = input.nextInt();
            int b = input.nextInt();
            union(a, b);
        }
        for (int i = 1; i < n + 1; i++) {
            System.out.println(find(i));
        }
    }
}

/* Sample Input/Output
 
    Input:
    6 4
    1 3
    3 4
    2 5
    5 6
    
    Output:
    1
    2
    1
    1
    2
    2
    
    Explanation:
    First line of input is n and e respectively.
    e line contain edges
    
    ith line of output corresponds to the set that i belongs to. (0 <= i <= n)
*/
